[Filters] Introduce GraphicsStyle and add it to GraphicsContext and GraphicsContextState

shallawa opened the pull request:
#### d6bd013ec6badee83d7af668ae1f5535714507f1
<pre>
[Filters] Introduce GraphicsStyle and add it to GraphicsContext and GraphicsContextState
<a href="https://bugs.webkit.org/show_bug.cgi?id=248197">https://bugs.webkit.org/show_bug.cgi?id=248197</a>
rdar://102591759

Reviewed by Cameron McCormack.

This will allow applying a GraphicsStyle to all the following drawing commands.
Some of FilterEffects can be expressed as GraphicsStyles. Applying a FilterEffect
using CoreGraphics will be done by setting GraphicsStyle to the GraphicsContext
before drawing the sourceImage of the Filter.

The plan is use transparency layers to control beginning, ending and nesting
GraphicsStyles.

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator&lt;&lt;):
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::stateChangeName):
(WebCore::GraphicsContextState::dump const):
* Source/WebCore/platform/graphics/GraphicsContextState.h:
(WebCore::GraphicsContextState::style const):
(WebCore::GraphicsContextState::setStyle):
(WebCore::GraphicsContextState::encode const):
(WebCore::GraphicsContextState::decode):
* Source/WebCore/platform/graphics/GraphicsStyle.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/GraphicsStyle.h: Added.
(WebCore::operator==):
(WebCore::GraphicsDropShadow::encode const):
(WebCore::GraphicsDropShadow::decode):
(WebCore::GraphicsGaussianBlur::encode const):
(WebCore::GraphicsGaussianBlur::decode):
(WebCore::GraphicsColorMatrix::encode const):
(WebCore::GraphicsColorMatrix::decode):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGStyle):
(WebCore::GraphicsContextCG::didUpdateState):

Canonical link: <a href="https://commits.webkit.org/256980@main">https://commits.webkit.org/256980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e48b9fdd026f154f230e1ba47cc438b1b88693b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6685 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106940 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167206 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7007 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35434 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103614 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103086 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5262 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84056 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32261 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87128 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75186 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88334 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/694 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83995 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/679 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21865 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28487 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5487 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44342 "Found 1 new test failure: fast/images/animated-heics-verify.html (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86821 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2372 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41218 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19509 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
```diff
@@ -190,9 +190,29 @@ struct CGShadowStyle {
+#if HAVE(CORE_GRAPHICS_STYLES)
+struct CGGaussianBlurStyle {
+    unsigned version;
+    CGFloat radius;   /* The blur's radius. */
```
Don't need to copy across this comment.

shallawa replied:
The comment was removed.

heycam commented on Source/WebCore/platform/graphics/GraphicsStyle.h:
```diff
@@ -0,0 +1,150 @@
+}
+
+struct GraphicsColorMatrix {
+    Vector<float> values;
```
Will this always be 20 values? If so, can we make this `std::array<float, 20>` instead?

weinig replied:
Also could ColorMatrix<4,5> class :).

shallawa replied:
`std::array<float, 20>` was used. I had to add `operator<<` for `TextStream` when the argument is `std::array`.

heycam commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1077,6 +1077,37 @@ void GraphicsContextCG::clearCGShadow()
+        },
+        [&] (const GraphicsColorMatrix& colorMatrix) -> RetainPtr<CGStyleRef> {
+            CGColorMatrixStyle colorMatrixStyle = { 1, { 0 } };
+            unsigned size = std::min(colorMatrix.values.size(), sizeof(colorMatrixStyle.matrix) / sizeof(colorMatrixStyle.matrix[0]));
```
Then we shouldn't need this.

heycam commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1106,6 +1137,12 @@ void GraphicsContextCG::didUpdateState(GraphicsContextState& state)
             setCGShadow(renderingMode(), state.dropShadow().offset, state.dropShadow().blurRadius, state.dropShadow().color, state.shadowsIgnoreTransforms());
             break;
 
+        case GraphicsContextState::Change::Style:
+#if HAVE(CORE_GRAPHICS_STYLES)
+            setCGStyle(context, state.style());
+#endif
+            break;
```
Should we be asserting in here in the non-`HAVE(CORE_GRAPHICS_STYLES)` state, or will we be unconditionally calling `GraphicsContextState::setStyle`, but skipping the existing accelerated / software implementations in the `HAVE(CORE_GRAPHICS_STYLES)` case?

shallawa replied:
`GraphicsContext::setStyle()` is not necessarily used by filters. For example we could implement `GraphicsContext::setDropShadow()` in terms of `GraphicsContext::setStyle()`.

shallawa committed:
[Filters] Introduce GraphicsStyle and add it to GraphicsContext and GraphicsContextState
https://bugs.webkit.org/show_bug.cgi?id=248197
rdar://102591759

Reviewed by Cameron McCormack.

This will allow applying a GraphicsStyle to all the following drawing commands.
Some of FilterEffects can be expressed as GraphicsStyles. Applying a FilterEffect
using CoreGraphics will be done by setting GraphicsStyle to the GraphicsContext
before drawing the sourceImage of the Filter.

The plan is use transparency layers to control beginning, ending and nesting
GraphicsStyles.

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator<<):
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::stateChangeName):
(WebCore::GraphicsContextState::dump const):
* Source/WebCore/platform/graphics/GraphicsContextState.h:
(WebCore::GraphicsContextState::style const):
(WebCore::GraphicsContextState::setStyle):
(WebCore::GraphicsContextState::encode const):
(WebCore::GraphicsContextState::decode):
* Source/WebCore/platform/graphics/GraphicsStyle.cpp: Added.
(WebCore::operator<<):
* Source/WebCore/platform/graphics/GraphicsStyle.h: Added.
(WebCore::operator==):
(WebCore::GraphicsDropShadow::encode const):
(WebCore::GraphicsDropShadow::decode):
(WebCore::GraphicsGaussianBlur::encode const):
(WebCore::GraphicsGaussianBlur::decode):
(WebCore::GraphicsColorMatrix::encode const):
(WebCore::GraphicsColorMatrix::decode):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGStyle):
(WebCore::GraphicsContextCG::didUpdateState):

Canonical link: https://commits.webkit.org/256980@main
